applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_linear_3D2N.cpp

```diff
@@ -0,0 +1,126 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:     BSD License
+//           license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Martin Fusseder, https://github.com/MFusseder
+//
+
+
+#include "adjoint_finite_difference_truss_element_linear_3D2N.h"
+#include "structural_mechanics_application_variables.h"
+#include "custom_response_functions/response_utilities/stress_response_definitions.h"
+
+
+namespace Kratos
+{
+
+AdjointFiniteDifferenceTrussElementLinear::AdjointFiniteDifferenceTrussElementLinear(Element::Pointer pPrimalElement)
+    : AdjointFiniteDifferenceTrussElement(pPrimalElement)
+{
+}
+
+AdjointFiniteDifferenceTrussElementLinear::~AdjointFiniteDifferenceTrussElementLinear()
+{
+}
+
+void AdjointFiniteDifferenceTrussElementLinear::Calculate(const Variable<Vector >& rVariable,
+                        Vector& rOutput,
+                        const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    if(rVariable == STRESS_ON_GP)
+    {
+        TracedStressType traced_stress_type = static_cast<TracedStressType>(this->GetValue(TRACED_STRESS_TYPE));
+
+        const SizeType  GP_num = (mpPrimalElement->GetGeometry().IntegrationPoints()).size();
+        if (rOutput.size() != GP_num)
+            rOutput.resize(GP_num, false);
+
+        switch (traced_stress_type)
+        {
+            case TracedStressType::FX:
+            {
+                std::vector< array_1d<double, 3 > > force_vector;
+                mpPrimalElement->GetValueOnIntegrationPoints(FORCE, force_vector, rCurrentProcessInfo);
+                for(IndexType i = 0; i < GP_num ; ++i)
+                    rOutput(i) = force_vector[i][0];
+                break;
+            }
+            default:
+                KRATOS_ERROR << "Invalid stress type! Stress type not supported for this element!" << std::endl;
+        }
+    }
+    else
+    {
+        rOutput.resize(1);
+        rOutput.clear();
+    }
+
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferenceTrussElementLinear::CalculateStressDisplacementDerivative(const Variable<Vector>& rStressVariable,
+                                    Matrix& rOutput, const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    TracedStressType traced_stress_type = static_cast<TracedStressType>(this->GetValue(TRACED_STRESS_TYPE));
+
+    if(traced_stress_type  == TracedStressType::FX)
+    {
+        // ensure that adjoint load is determined without influence of pre-stress
+        // pre-stress does not cancel out when computing this derivative with unit-displacements!
+        Properties::Pointer p_global_properties = mpPrimalElement->pGetProperties();
+
+        Properties::Pointer p_local_property(Kratos::make_shared<Properties>(Properties(*p_global_properties)));
```

Comment:
this seems strange, @msandre could you please check this?

Reply:
Copy constructor nested in a shared pointer creation -  should be fine. I think this is used also in other adjoint finite differencing elements. If there is something wrong with this we need to change it there too.

Reply:
@armingeiser you are right this is also used within the adjoint finite differencing base element.

Reply:
It looks like you are creating an unnecessary temporary with `Properties(*p_global_properties)`. I.e., you are calling the copy constructor twice. Why not just do `Properties::Pointer p_local_property(Kratos::make_shared<Properties>(*p_global_properties));`?